src/protocol/mod.rs

```diff
@@ -93,6 +93,15 @@ impl PartialEq for UniqueStepId {
 impl Eq for UniqueStepId {}
 
 impl UniqueStepId {
+    #[must_use]
+    pub fn from_step_id(step: &Self) -> Self {
+        Self {
+            id: step.id.clone(),
+            #[cfg(debug_assertions)]
+            used: Arc::new(Mutex::new(HashSet::new())),
+        }
+    }
```

Comment:
Wait... am I understanding this correctly?

It looks like every time you call `ctx.bind(record_id)` it creates a brand-new, empty hash set, with associated wrapping Mutex. That seems wrong to me. Isn't the idea that all of the contexts would share a pointer to the *same* hash set?

Reply:
For this implementation yes it is creating a hash set per record ID. However, since opt-level=3 doesnt look at debug_assertions, do you think it is ok to leave it like this for now until we get to a better solution?

Reply:
that's my code and it looks funky I agree. We need to create a brand new hash set to make sure we can narrow to the same step more than once. One alternative to that is to keep record id in the hashset which is bad. This code is just to unblock testing sort and we need a better long term solution for sure #135

Reply:
Per our VC discussion today I get it: each record-bound-context is going to be a totally independent entity, only validating that it isn't narrowed to the same exact step multiple times. This provides no protection against PRSS re-use - and that will be addressed elsewhere, e.g. #208